File: lib/linux/ip/addr.rb
require "linux/ip/addr/version"

module Linux
  module Ip
    module Addr

      class Interface
        attr_accessor :mac_address
        attr_reader :ips, :name
        def initialize(name)
          @name = name
          @ips = []
        end
        def add_ip(ip)
          @ips << ip
        end
      end


      class IpAddr
        attr_reader :interfaces
        def initialize
          @interfaces = []
        end
        def length
          interfaces.length
        end
        def find(name)
          interfaces.find { |i| i.name == name }
        end
      end

      #23: br202: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default
      #    link/ether d0:27:88:d1:01:6d brd ff:ff:ff:ff:ff:ff
      #    inet 10.10.103.11/22 brd 10.10.103.255 scope global br202
      #       valid_lft forever preferred_lft forever
      #    inet6 fe80::84c6:95ff:feb7:d76d/64 scope link
      #       valid_lft forever preferred_lft forever
      def self.parse_from_lines(lines)
        ip_addr = IpAddr.new
        iface = nil
        lines.each do |line|
            line = line.strip.chomp
            if line =~ /^\d+:\s+(\S+):\s+(.*)$/
              iface = Interface.new($1.split('@')[0])
              ip_addr.interfaces << iface
            elsif iface
              if line =~ /^\s*link\/ether\s+([a-f0-9:]+)\s+.*$/
                iface.mac_address = $1
              elsif line =~ /^\s*inet[6]*\s+([0-9a-f\:\.\/]+)\s+.*$/
                iface.add_ip($1)
              end
            end
        end
        ip_addr
      end

      def self.parse
        parse_from_lines(IO.popen("ip addr show").read.lines)
      end

    end
  end
end

File: rust/src/linux/ip/addr.rs

use std::collections::LinkedList;

struct Interface {
    mac_address: String,
    name: String,
    ips: Vec
}
impl Interface {
    fn add_ip(ip: &str) {
        ips.push_back(i);
        self
    }
    fn as_commands(direction) {
        ips.iter.map(|ip|
          format!("ip addr {} {} dev {}", direction, ip, self.name)
        )
    }
}

struct IpAddr {
  interfaces: Vec<Interface>
}

impl IpAddr {
  fn length() {
    interfaces.len()
  }
  fn find(name: &str) {
    interfaces.iter().find( |i| i.name == name )
  }
  fn as_commands(direction: &str) {
    let mut ret = Vec<String>::new
    for interface in interfaces {
        ret.append(interface.as_commands(direction))
    }
    ret
  }
  fn execute_add() {
    system as_commands("add")
  }
  fn execute_del() {
    system as_commands("del")
  }
}


impl Addr {
    pub fn parse_from_lines(lines: &str) -> IpAddr {
      let ip_addr = IpAddr::new {
        ips: Vec::new()
      }
      let re_crnl = Regex::new(r"[\s\n\r]+").unwrap();
      let re_line = Regex::new(r"^\d+:\s+(\S+):\s+(.*)$").unwrap();
      let re_mac_address = Regex::new(r"\s*link\/ether\s+([a-f0-9:]+)\s+.*").unwrap();
      let re_ip_addr = Regex.new(r"\s*inet[6]*\s+([0-9a-f\:\.\/]+)\s+.*").unwrap();
      let mut op_iface: Option<Interface> = None;
      for line in RegexSplits(re_crnl, lines) {
          if re_line.is_match(line) {
            op_iface = Interface.new($1.split('@')[0])
            ip_addr.interfaces << iface
        } else {
            match op_iface {
                Some(iface) => {
                    for cap in re_mac_address.captures_iter(line) {
                        iface.mac_address = cap.at(1).unwrap();
                    }
                    for cap in re_ip_addr.captures_iter(line) {
                        iface.ips.push_back(cap.at(1).unwrap());
                    }
                }
                None => {}
            }
        }
      }
      return ip_addr;
    }

    pub fn parse() {
        let out = Command::new("ip").arg("addr").arg("show").output().unwrap();
        parse_from_lines(String::from_utf8_lossy(&out.stdout))
    }
}
